runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -12418,6 +12418,905 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+    * indices are being passed in as ints, we must ensure that their upper 32 bits are not garbage.
+    */
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::extsw, node, resultReg, fromIndexReg);
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::extsw, node, subStringLenReg, strCountReg);
+
+   if (node->getChild(3)->getReferenceCount() == 1)
+      {
+      srm->donateScratchRegister(strCountReg);
+      }
+
+   if (node->getChild(4)->getReferenceCount() == 1)
+      {
+      srm->donateScratchRegister(fromIndexReg);
+      }
+
+   /* Under UTF16, the indices are doubled since offset and size in bytes are needed later. */
+   if (!isLatin1)
+      {
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, resultReg, resultReg, resultReg);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, subStringLenReg, subStringLenReg, subStringLenReg);
+      }
+
+   /*
+    * Determine the address of the first byte to read either by loading from dataAddr or adding the header size.
+    * This is followed by adding in the offset.
+    */
+   TR::Register *subStringAddressReg = srm->findOrCreateScratchRegister();
+#if defined(J9VM_GC_SPARSE_HEAP_ALLOCATION)
+   if (TR::Compiler->om.isOffHeapAllocationEnabled())
+      {
+      generateTrg1MemInstruction(
+         cg, TR::InstOpCode::ld, node, arrayAddressReg,
+         TR::MemoryReference::createWithDisplacement(
+            cg, arrayObjReg, TR::Compiler->om.offsetOfContiguousDataAddrField(), 8)
+         );
+      generateTrg1MemInstruction(
+         cg, TR::InstOpCode::ld, node, subStringAddressReg,
+         TR::MemoryReference::createWithDisplacement(
+            cg, subStringObjReg, TR::Compiler->om.offsetOfContiguousDataAddrField(), 8)
+         );
+      }
+   else
+#endif /* J9VM_GC_SPARSE_HEAP_ALLOCATION */
+      {
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, arrayAddressReg, arrayObjReg,
+                                     TR::Compiler->om.contiguousArrayHeaderSizeInBytes());
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, subStringAddressReg, subStringObjReg,
+                                     TR::Compiler->om.contiguousArrayHeaderSizeInBytes());
+      }
+
+   if (node->getChild(0)->getReferenceCount() == 1)
+      {
+      srm->donateScratchRegister(arrayObjReg);
+      }
+
+   if (node->getChild(2)->getReferenceCount() == 1)
+      {
+      srm->donateScratchRegister(subStringObjReg);
+      }
+
+   TR::Register *currentAddressReg = srm->findOrCreateScratchRegister();
+
+   /* resultReg currently holds the starting offset. currentAddressReg will hold the address of the first byte of the main string to look at. */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, currentAddressReg, arrayAddressReg, resultReg);
+
+   TR::Register *firstCharReg = srm->findOrCreateScratchRegister();
+   TR::Register *firstCharSubStrReg = srm->findOrCreateScratchRegister();
```

Comment:
i am wondering if you can optimize the use of these many registers. at this point, 13 or 14 GPRs are needed already.

Reply:
My current implementation uses:
11-16 GPR (Depends on how many input Regs have ref count of 0 and can be reused. Saves up to 5 registers)
7 VRF
2 CCR

For comparison, the Arm implementation:
https://github.com/eclipse-openj9/openj9/blob/45733dc106ac9a52570239011e90eb61ae01fd74/runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp#L6672-L6701
This uses:
11-13 GPR (Depends on how many input Regs have ref count of 0 and can be reused. Saves up to 2 registers)
3 VRF